doc/codegen/modular_codegen_design.md

```diff
@@ -0,0 +1,561 @@
+- **Input**: Raw SDFG
+- **Output**: SDFG with inferred types, pipeline_results["type_info"]
+- **Current Location**: `infer_types.py` functions
+
+#### 2. **LibraryExpansionPass**
+- **Purpose**: Expand all library nodes that haven't been expanded
+- **Input**: Type-inferred SDFG
+- **Output**: SDFG with expanded library nodes
+- **Current Location**: `sdfg.expand_library_nodes()`
+
+#### 3. **MetadataCollectionPass**
+- **Purpose**: Collect free symbols, argument lists, constants, shared transients
+- **Input**: Expanded SDFG
+- **Output**: pipeline_results["metadata"] = {symbols, arglist, constants, shared_transients}
+- **Current Location**: `DaCeCodeGenerator.__init__()`
+
+#### 4. **AllocationAnalysisPass**
+- **Purpose**: Determine allocation lifetimes and scopes for all data containers
+- **Input**: SDFG with metadata
+- **Output**: pipeline_results["allocation_info"] = {to_allocate, where_allocated}
+- **Current Location**: `DaCeCodeGenerator.determine_allocation_lifetime()`
+
+#### 5. **ControlFlowAnalysisPass**
+- **Purpose**: Extract structured control flow from state machines
+- **Input**: SDFG
+- **Output**: pipeline_results["control_flow_tree"]
+- **Current Location**: `control_flow.py` functions
+
+#### 6. **TargetAnalysisPass**
+- **Purpose**: Identify required code generation targets and dispatch routes
+- **Input**: SDFG with metadata
+- **Output**: pipeline_results["targets"] = {required_targets, dispatch_info}
+- **Current Location**: `_get_codegen_targets()`
+
+### Phase 2: Transformation Passes
+
+#### 7. **CopyToMapPass**
+- **Purpose**: Convert complex memory copies to Map nodes where needed
+- **Input**: SDFG with targets identified
+- **Output**: SDFG with transformed copies
+- **Current Location**: `cuda.py` preprocessing, various target preprocessors
+- **Applies To**: GPU strided copies, FPGA transfers
+
+#### 8. **StreamAssignmentPass** (GPU-specific)
+- **Purpose**: Assign CUDA streams for concurrent execution
+- **Input**: GPU-targeted SDFG
+- **Output**: SDFG with stream assignments, pipeline_results["stream_info"]
+- **Current Location**: Embedded in CUDA code generation
+
+#### 9. **TaskletLanguageLoweringPass**
+- **Purpose**: Convert Python/generic tasklets to target language (C++/CUDA/etc.)
+- **Input**: SDFG with tasklets
+- **Output**: SDFG with lowered tasklets
+- **Current Location**: Distributed across target generators
+
+### Phase 3: Code Generation Passes
+
+#### 10. **StateStructCreationPass**
+- **Purpose**: Generate state struct definitions for persistent data
+- **Input**: SDFG with allocation info
+- **Output**: pipeline_results["state_struct"] = {struct_def, struct_init}
+- **Current Location**: `DaCeCodeGenerator.generate_code()`
+
+#### 11. **AllocationCodePass**
+- **Purpose**: Generate allocation/deallocation code for each scope
+- **Input**: SDFG with allocation analysis
+- **Output**: pipeline_results["allocation_code"] = {alloc_code, dealloc_code}
```

Comment:
The output here should not be allocation code, but should rather still modify the SDFG. Adding allocation and deallocation tasklets to the graph seems simplest. Additionally, this should happen earlier in the pipeline.